liblangutil/ParserBase.cpp

```diff
@@ -57,30 +57,79 @@ Token ParserBase::advance()
 	return m_scanner->next();
 }
 
+std::string ParserBase::tokenName(Token _token)
+{
+	if (_token == Token::Identifier)
+		return "identifier";
+	else if (_token == Token::EOS)
+		return "end of source";
+	else if (TokenTraits::isReservedKeyword(_token))
+		return "reserved keyword '" + TokenTraits::friendlyName(_token) + "'";
+	else if (TokenTraits::isElementaryTypeName(_token)) //for the sake of accuracy in reporting
+	{
+		ElementaryTypeNameToken elemTypeName = m_scanner->currentElementaryTypeNameToken();
+		return "'" + elemTypeName.toString() + "'";
+	}
+	else
+		return "'" + TokenTraits::friendlyName(_token) + "'";
+};
+
 void ParserBase::expectToken(Token _value, bool _advance)
 {
 	Token tok = m_scanner->currentToken();
 	if (tok != _value)
 	{
-		auto tokenName = [this](Token _token)
+		std::string const expectToken = ParserBase::tokenName(_value);
+		if (m_parserErrorRecovery)
+			parserError("Expected " + expectToken + " but got " + tokenName(tok));
+		else
+			fatalParserError("Expected " + expectToken + " but got " + tokenName(tok));
+		// Do not advance so that recovery can sync or make use of the current token. This is especially useful if the expected token
+		// is the only one that is missing and is at the end of a construct.
+		// "{ ... ; }" is such an example.
+		//        ^
+		_advance = false;
+	}
+	if (_advance)
+		m_scanner->next();
+}
+
+void ParserBase::expectTokenOrConsumeUntil(Token _value, char const *_lhs, bool _advance)
```

Comment:
```suggestion
void ParserBase::expectTokenOrConsumeUntil(Token _value, string _currentNode, bool _advance)
```

(what does `lhs` stand for?)

Reply:
left-hand-side. In my experience it is often used in referring to grammar rules just as it is in programming language assignment statements.

However if you would like to suggest another name, please do!

Reply:
`_currentNode`?